### DISCUSSION TITLE
운영체제 - 메인 메모리

### DISCUSSION THREAD

**Discussion by seongeun42** [Shortudy-10th/tech-for-developer#147] [질의 응답]
각자 생각나는 질문&답변 같이 달기

**Comment by zini9188** [comment]
## Q. 메인 메모리란 무엇인가요?

메인 메모리란 CPU가 데이터와 명령어를 저장하고 검색하기 위해 직접 접근할 수 있는 임시 저장 공간으로 램이라고도 불립니다.

**Reply by watty-exclamationMark** [reply]
## Q. 메모리 영역에 대해 말해주세요.
메모리는 코드(텍스트) 영역, 데이터 영역, 힙 영역, 스택 영역으로 이루어져 있습니다.

**Reply by watty-exclamationMark** [reply]
## Q. 스택 영역과 힙 영역에 대한 차이점을 말해주세요.
스택 영역은 지역 변수, 매개 변수를 저장하는 곳이고 힙 영역은 동적으로 생성된 변수를 저장하는 곳입니다.
스택은 사용자가 직접 관리할 필요가 없으나 힙 영역은 사용자에 의해 메모리 공간이 할당되고 해제됩니다.

**Comment by Jihn0118** [comment]
## ❓️ 캐시 메모리는 무엇이고 왜 쓰는 건가요?

💬 답변
캐시 메모리는 CPU와 메인 메모리 사이에 위치한 고속의 작은 용량을 가진 메모리이고, 데이터와 명령어의 빈번한 액세스를 빠르게 처리하기 위한 임시 저장 공간입니다.

캐시 메모리는 주로 메인 메모리와 CPU 간의 속도 차이를 완화하고 프로세서의 성능을 향상 시키기 위해 사용됩니다.

**Comment by seongeun42** [comment]
## Q. 메모리의 절대 주소와 상대 주소에 대해 설명해주세요.
절대 주소란 변수 등이 실제로 메모리 내에 위치하는 물리 주소를 의미하고, 상대 주소는 실제 위치를 무시하고 항상 0부터 시작하는 논리 주소 공간 속의 주소를 의미합니다.

**Comment by seongeun42** [comment]
## Q. 메모리 배치 정책에는 어떤 것들이 있나요?

프로세스 크기에 따라 메모리를 나누는 가변 분할 방식과 크기와 상관없이 메모리를 일정한 크기로 나누는 고정 분할 방식이 있습니다.

**Reply by seongeun42** [reply]
## Q. 고정 분할 방식이란 무엇인가요?

메모리를 같은 크기로 나누고, 그보다 큰 프로세스가 메모리에 올라오면 프로세스를 작은 단위로 쪼개 여러 조각으로 나누어서 배치하는 방식입니다. 한 프로세스가 분산되어 배치되기 때문에 비연속 메모리 할당이라고도 합니다.

**Reply by seongeun42** [reply]
## Q. 가변 분할 방식이란 무엇인가요?

올라오는 프로세스의 크기에 맞춰 메모리를 분할하고 프로세스를 배치하는 방식입니다. 한 프로세스가 쪼개지지 않고 연속된 공간에 배치되기 때문에 연속 메모리 할당이라고도 합니다.

**Comment by sikutisa** [comment]
## Q. Dynamic Loading이란 무엇인가요?
프로세스가 시작될 때 프로세스의 주소 공간 전체를 메모리에 올려놓는 것이 아니라, 필요한 루틴이 호출될 때 해당 루틴만을 메모리에 올려 메모리를 효율적으로 사용하는 것 입니다.

**Comment by sikutisa** [comment]
## Q. Variable Partition이란 무엇인가요?
초기에는 메모리 전체를 하나의 영역으로 관리하고, 프로세스를 처리하는 과정에서 메모리 공간을 동적으로 할당하여 배정하는 방식입니다. 이때, 분할된 공간의 크기는 각 프로세스가 필요로 하는 영역의 크기만큼 분할합니다.

**Reply by sikutisa** [reply]
## Q. Dynamic Storage Allocation의 메모리 할당 전략에 대해 설명해주세요.
First-Fit은 빈 공간(hole)들 중 메모리 할당이 가능한 첫번째 공간을 할당하는 방식입니다. Best-Fit은 빈 공간들 중 메모리 할당이 가능한 가장 작은 공간을 할당하는 방식입니다. Worst-Fit은 빈 공간들 중 메모리 할당이 가능한 가장 큰 공간을 할당하는 방식입니다.

**Comment by sikutisa** [comment]
## Q. Fragmentation이란 무엇인가요
메모리가 여러번 할당되고 반납되는 과정에서 전체 메모리 공간이 작은 공간 여러개로 쪼개져, 사용 가능한 메모리가 충분함에도 불구하고 메모리 할당이 불가능해지는 상태입니다.

**Reply by sikutisa** [reply]
## Q. Fragmentation의 종류에 대해 설명해주세요.
내부 단편화는 메모리 내 사용자 영역이 실행 프로그램보다 커서 프로그램의 사용 공간을 할당한 후 사용되지 않고 남아있는 공간을 의미합니다.
외부 단편화는 사용자 영역보다 실행 프로그램이 커서 프로그램이 할당되지 못하고 사용되지 않은 채 남아있는 공간을 의미합니다.

**Comment by aeroej** [comment]
내부 단편화와 외부 단편화는 무엇인가요?

**Reply by aeroej** [reply]
![IMG_0123](https://github.com/Shortudy-10th/tech-for-developer/assets/39663810/641d5d8f-857f-4882-a378-4cf1d9260088)

내부단편화는 메모리에 할당할 수 있지만 그 과정에서 메모리 낭비가 발생하는 현상입니다.
외부단편화는 남은 공간을 합치면 충분히 메모리에 할당할 수 있음에도, 남은 공간이 연속적이지 않아 할당할 수 없는 현상입니다.
